torch/_custom_op/autograd.py

```diff
@@ -0,0 +1,222 @@
+                schema, pytree.tree_map(lambda arg: type(arg), args))
+
+            save_for_backward_fn_inputs = namedtuple_args(schema, args)
+            to_save = save_for_backward_fn(save_for_backward_fn_inputs, output)
+
+            save_pytree_for_backward(ctx, (to_save, args_info))
+
+            # Output must be one or more Tensors, no TensorList (yet)
+            if output_differentiability is not None:
+                if isinstance(output, tuple):
+                    assert len(output_differentiability) == len(output)
+                    for differentiable, out in zip(output_differentiability, output):
+                        if not differentiable:
+                            ctx.mark_non_differentiable(out)
+                else:
+                    assert len(output_differentiability) == 1
+                    if not output_differentiability[0]:
+                        ctx.mark_non_differentiable(output)
+
+            return output
+
+        def backward(ctx, *grads):
+            saved, args_info = unpack_saved(ctx)
+            # There is nothing on the ctx object for now, it is just there so
+            # that we can add additional things in the future.
+            inner_ctx = object()
+            grad_inputs_dict = backward_fn(inner_ctx, saved, *grads)
+
+            # Massage the grad_inputs_dict to a form acceptable by
+            # autograd.Function.
+            validate_grad_inputs_dict(grad_inputs_dict, forward_op, args_info)
+            return grad_inputs_dict_to_flat_tuple(grad_inputs_dict, args_info)
+
+        generated_cls = gen_autograd_function(
+            forward_op._opname + 'CustomOp', forward, backward)
+
+        return generated_cls.apply(*flat_args)
+    return apply
+
+
+def gen_autograd_function(name, forward, backward):
+    generated_cls = type(
+        name,
+        (torch.autograd.Function,),
+        {
+            'forward': staticmethod(forward),
+            'backward': staticmethod(backward),
+        }
+    )
+    return generated_cls
+
+
+def namedtuple_args(schema, args):
+    assert isinstance(args, tuple)
+    attribs = [arg.name for arg in schema.arguments.flat_all]
+    name = str(schema.name) + "Args"
+    tuple_cls = namedtuple(name, attribs)
+    return tuple_cls(*args)
+
+
+def validate_grad_inputs_dict(grad_inputs_dict, forward_op, args_info):
+    def error(what):
+        backward = forward_op._get_impl('backward')
+        raise RuntimeError(
+            f"In the backward function defined for {forward_op} at "
+            f"{backward.location} using the CustomOp API, {what}")
+
+    if not isinstance(grad_inputs_dict, dict):
+        error(f"expected the output of the backward function to be a dict but "
+              f"got {type(grad_inputs_dict)}")
+
+    expected_keys = {arg.name for arg in forward_op._schema.arguments.flat_all
+                     if arg.type.is_tensor_like()}
+    actual_keys = {k for k, _ in grad_inputs_dict.items()}
```

Comment:
grad_inputs_dict.keys() ?
You can compare the result of this with the dict above or a set.

Reply:
I was a bit scared that the type of `dict.keys()` is not actually a dict or a set, but I'll try it out

Reply:
Yes, this works